### PR TITLE
Expose foreign keys via feature flag

### PR DESCRIPTION
**Pull request by errorhandler**
Following the discussion in #869 I've gone ahead and given 1. and 2. a shot, with 3. coming in a follow-up PR. I've added a basic test case and put this feature behind a feature flag.

Any feedback is greatly appreciated! Especially in the general approach.

### CODE REVIEW

File: entc/gen/feature.go [hunk 1]
```diff
@@ -43,11 +43,20 @@ var (
 		},
 	}
 
+	FeatureExposeFKs = Feature{
+		Name:        "exposefks",
```
Comment (errorhandler) [inline entc/gen/feature.go:47]: Any better names would be greatly appreciated.

File: entc/gen/template/dialect/sql/create.tmpl [hunk 2]
```diff
@@ -67,7 +67,16 @@ func ({{ $receiver }} *{{ $builder }}) createSpec() (*{{ $.Name }}, *sqlgraph.Cr
 			_spec.Edges = append(_spec.Edges, edge)
 		}
 	{{- end }}
-	return _node, _spec
+	{{- if ($.FeatureEnabled "exposefks") }}
+	{{- range $fk := $.ForeignKeys }}
+	{{- with $e := $.EdgeForForeignKey $fk }}
+		if id, ok := {{ $mutation }}.{{ $e.StructField }}ID(); ok {
+			_node.{{ $fk.Field.StructField }} = {{ if $fk.Field.Nillable }}&{{ end }}id
```
Comment (errorhandler) [inline entc/gen/template/dialect/sql/create.tmpl:74]: If we're exposing the foreign keys, set the field if the foreign key was set in the mutation.

File: entc/gen/template/dialect/sql/decode.tmpl [hunk 3]
```diff
@@ -67,8 +67,12 @@ func ({{ $receiver }} *{{ $.Name }}) assignValues(values ...interface{}) error {
 						if value, ok := values[{{ $i }}].(*sql.NullInt64); !ok {
 							return fmt.Errorf("unexpected type %T for edge-field {{ $f.Name}}", value)
 						} else if value.Valid {
-							{{ $receiver }}.{{ $f.Name }} = new({{ $f.Type }})
-							*{{ $receiver }}.{{ $f.Name }} = {{ $f.Type }}(value.Int64)
+							{{- if $f.Nillable }}
```
Comment (errorhandler) [inline entc/gen/template/dialect/sql/decode.tmpl:70]: Handle both pointers and value foreign key fields.

File: entc/gen/template/dialect/sql/query.tmpl [hunk 4]
```diff
@@ -7,7 +7,9 @@ in the LICENSE file in the root directory of this source tree.
 {{/* Additional fields for the builder. */}}
 {{ define "dialect/sql/query/fields" }}
 	{{- with $.ForeignKeys }}
-		withFKs bool
+		{{- if not ($.FeatureEnabled "exposefks") }}
+			withFKs bool
```
Comment (errorhandler) [inline entc/gen/template/dialect/sql/query.tmpl:11]: This is always true if the FF is enabled, so don't generated it if the FF is on.

File: entc/gen/template/dialect/sql/query.tmpl [hunk 5]
```diff
@@ -40,15 +45,22 @@ func ({{ $receiver }} *{{ $builder }}) sqlAll(ctx context.Context) ([]*{{ $.Name
 			if withFKs {
 				_spec.Node.Columns = append(_spec.Node.Columns, {{ $.Package }}.ForeignKeys...)
 			}
+			{{- else }}
+			_spec.Node.Columns = append(_spec.Node.Columns, {{ $.Package }}.ForeignKeys...)
```
Comment (errorhandler) [inline entc/gen/template/dialect/sql/query.tmpl:49]: Always load the FK if the FF is set.

**Comment by alexsn** [comment]
I dislike APIs where there are multiple ways to do the same thing. Currently if we want the edge id of an object be must do `<Object>.Query<Edge>.OnlyID()` and I'd prefer if this will remain the only way. So I'd suggest the FK remain unexported and the `.OnlyID` query finisher will check if we can fetch the ID without db call (i.e. see if it's cached on those unexported fields).

**Comment by a8m** [comment]
> Following the discussion in #869 I've gone ahead and given 1. and 2. a shot, with 3. coming in a follow-up PR. I've added a basic test case and put this feature behind a feature flag.
> 
> Any feedback is greatly appreciated! Especially in the general approach.

I'd appreciate if you can add a summary how do you expect/plan the API to look like. Currently, I'm using custom templates to implement an API with similar functionality:
```go
client.User.
    Query().
    WithEdgeIDs() // We can rename it to "WithFKs".
    All(ctx)

// Then, use it as follows:
id, err := client.CardID()
```

**Comment by a8m** [comment]
Also, it's part of a wider change that will allow selecting specific node fields when querying data - much more efficient in some cases (GraphQL in my case).

**Comment by G-Bruin** [comment]
@a8m When will this new feature be available？

**Comment by a8m** [comment]
> Also, it's part of a wider change that will allow selecting specific node fields when querying data - much more efficient in some cases (GraphQL in my case).

This change should be added with https://github.com/facebook/ent/pull/1075, and we can rebase the work here on top of it. 

Thanks for working on it @errorhandler

**Comment by a8m** [comment]
> > Also, it's part of a wider change that will allow selecting specific node fields when querying data - much more efficient in some cases (GraphQL in my case).
> 
> This change should be added with #1075, and we can rebase the work here on top of it.
> 
> Thanks for working on it @errorhandler

Hey @errorhandler, any update on this PR? Do you need help with this? 
Let me know if you don't have time to finish it, and I can continue it. Thanks 😃

**Comment by errorhandler** [comment]
> > > Also, it's part of a wider change that will allow selecting specific node fields when querying data - much more efficient in some cases (GraphQL in my case).
> > 
> > 
> > This change should be added with #1075, and we can rebase the work here on top of it.
> > Thanks for working on it @errorhandler
> 
> Hey @errorhandler, any update on this PR? Do you need help with this?
> Let me know if you don't have time to finish it, and I can continue it. Thanks 😃

Hi @a8m! I was going to finish it up after New Years, however I've had a death in the family which means I have not be able to work on this. I'll have to pass this over to you I'm afraid.

Thanks for all your hard work on ent @a8m!

**Comment by a8m** [comment]
> Hi @a8m! I was going to finish it up after New Years, however I've had a death in the family which means I have not be able to work on this. I'll have to pass this over to you I'm afraid.
> 
> Thanks for all your hard work on ent @a8m!

Hey @errorhandler, it made me sad to read it yesterday 😢. I'm really sorry for your loss and hope you and your family are safe and healthy.

Thanks for taking the time to work on this. I really appreciate it 🙏

**Comment by a8m** [comment]
Closing as this was addressed with #1213 

Thanks!

**Pull request closed without merging**